### PR TITLE
Internal: Editor V2 - i18n [ED-9205]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
+			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
+				const translationsCallExpression = this.getTranslationsCallExpression( compilation );
+
+				this.addTranslationsCallExpressionToAssets( compilation, translationsCallExpression );
+			} );
+		} );
+	}
+
+	getTranslationsCallExpression( compilation ) {
```
Comment (StyleShit) [inline packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:34]:
```suggestion
	getTranslationCallExpressions( compilation ) {
```

also others

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: core/editor/config-providers/editor-common-assets.php [hunk 2]
```diff
@@ -110,6 +110,9 @@ public static function get_script_configs() {
 					'react',
 					'react-dom',
 				],
+				'translations' => [
+					'active' => true,
+				],
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/config-providers/editor-common-assets.php:115]: Maybe

File: core/editor/config-providers/editor-v2-config-provider.php [hunk 3]
```diff
@@ -14,29 +14,45 @@ public function get_script_configs() {
...
 				],
 				[
 					'handle' => 'elementor-packages-top-bar',
-					'src' => '{{ASSETS_URL}}/js/packages/top-bar{{MIN_SUFFIX}}.js',
-					'deps' => [ 'react', 'elementor-packages-editor', 'elementor-packages-ui' ],
+					'src' => '{{ASSETS_URL}}js/packages/top-bar{{MIN_SUFFIX}}.js',
+					'deps' => [ 'react', 'elementor-packages-editor', 'elementor-packages-ui', 'wp-i18n' ],
+					'translations' => [
+						'active' => true,
+						'path_suffix' => 'strings',
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/config-providers/editor-v2-config-provider.php:30]: Done

File: core/editor/editor-loader.php [hunk 4]
```diff
@@ -90,20 +104,47 @@ public function print_root_template() {
...
 		$additional_defaults = 'style' === $type ?
 			[ 'media' => 'all' ] :
-			[ 'in_footer' => true ];
-
-		$default = array_merge( [
+			[
+				'in_footer' => true,
+				'translations' => [
+					'active' => false,
+					'domain' => 'elementor',
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:142]: We should talk about it

File: core/editor/editor-loader.php [hunk 5]
```diff
@@ -135,7 +176,39 @@ private function normalize_asset_configs( array $script_configs, $type ) {
 				}
 
 				return $config;
-			} )
-			->all();
+			} );
+	}
+
+	/**
+	 * The name of the translation file could be suffix with `strings` to support loading translations
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:183]: Done

File: core/editor/editor-loader.php [hunk 6]
```diff
@@ -135,7 +176,39 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
+	 * for lazy loaded scripts.
+	 *
+	 * Want to go deeper:
+	 * @link https://developer.wordpress.com/2022/01/06/wordpress-plugin-i18n-webpack-and-composer/
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:187]: Done

File: core/editor/editor-loader.php [hunk 7]
```diff
@@ -135,7 +176,39 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
+
+		$should_suffix_path = $script_config &&
+			$script_config['translations']['active'] &&
+			$script_config['translations']['path_suffix'];
+
+		if ( ! $should_suffix_path ) {
+			return $relative_path;
+		}
+
+		// Translations are always based on the unminified filename.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:208]: Done

File: core/editor/editor-loader.php [hunk 8]
```diff
@@ -135,7 +176,39 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
+		$substr_offset = substr( $relative_path, -7 ) === '.min.js' ? -7 : -3;
+		$relative_path_without_ext = substr( $relative_path, 0, $substr_offset );
+
+		return "{$relative_path_without_ext}.{$script_config['translations']['path_suffix']}.js";
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:212]: Done

File: packages/.eslintrc.js [hunk 9]
```diff
@@ -30,11 +27,12 @@ module.exports = {
 			},
 		],
 
-		// Disable import/named rule, TS will handle it.
+		// Disable rules, TS will handle it.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/.eslintrc.js:30]: Done

File: tests/phpunit/elementor/core/editor/test-editor-loader.php [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	public function test_load_script_translations() {
+		// Arrange
+		global $wp_scripts;
+
+		$this->mock_config_provider->method( 'get_script_configs' )
+			->willReturn( [
+				 [
+				     'handle' => 'script-with-translations',
+				     'src' => 'http://localhost',
+					 'deps' => [ 'some-dep' ],
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/editor/test-editor-loader.php:30]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js (hunk 1 above)
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:34]: Done

File: tests/phpunit/elementor/core/editor/test-editor-loader.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+			] );
+
+		$loader = new Editor_Loader( $this->mock_config_provider );
+		$loader->register_scripts();
+
+		// Act
+		$loader->load_scripts_translations();
+
+		// Assert
+		$this->assertEquals( $wp_scripts->registered['script-with-translations']->textdomain, 'elementor' );
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/editor/test-editor-loader.php:112]: Done

File: packages/webpack.common.js [hunk 12]
```diff
@@ -3,6 +3,7 @@ const { readdirSync } = require( 'fs' );
 
 const globalObjectKey = '__UNSTABLE__elementorPackages';
 
+// All our internal packages
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/webpack.common.js]: Done

**Review by StyleShit — DISMISSED** [review]

File: core/editor/editor-loader.php [hunk 13]
```diff
@@ -135,7 +183,42 @@ private function normalize_asset_configs( array $script_configs, $type ) {
 				}
 
 				return $config;
-			} )
-			->all();
+			} );
+	}
+
+	/**
+	 * The name of the translation file could be suffixed with a translation's identifier (usually ".strings")
```
Comment (rotemee) [inline core/editor/editor-loader.php:190]:
```suggestion
	 * The name of the translation file should be suffixed with a translation's identifier (usually ".strings")
```

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+				return;
+			}
+
+			const mainFilePath = compilation.options.entry[ id ].import[ 0 ];
+
+			const assetFilename = compilation
+				.getPath( '[file]', { filename: chunkJSFile } )
+				.replace( /(\.min)?\.js$/i, '.strings.js' );
+
+			// Add source and file into compilation for webpack to output.
```
Comment (rotemee) [inline packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:86]: source and file?

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+						return;
+					}
+
+					const mainEntryFile = this.findMainModuleOfEntry( subModule, compilation );
+
+					if ( ! translationCallExpressions.has( mainEntryFile ) ) {
+						translationCallExpressions.set( mainEntryFile, new Set() );
+					}
+
+					this.translationsRegexps.forEach( ( regexp ) => {
```
Comment (rotemee) [inline packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:59]: Pleas explain.

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+						} );
+					} );
+				} );
+			} );
+		} );
+
+		return translationCallExpressions;
+	}
+
+	addTranslationCallExpressionsToAssets( compilation, translationCallExpressions ) {
```
Comment (rotemee) [inline packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:71]: Please explain.

File: packages/webpack.prod.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+	optimization: {
+		minimize: true,
+	},
+	output: {
+		filename: '[name].min.js',
+	},
+	plugins: [
+		new ReadableJsAssetsWebpackPlugin(),
+		new ExtractI18nExpressionsWebpackPlugin( {
+			translationsRegexps: [ /\b_(?:_|n|nx|x)\(.+,\s*(?<c>['"`])[\w-]+\k<c>\)/ ],
```
Comment (rotemee) [inline packages/webpack.prod.js:18]: Consider adding a comment with an example of the Regex.

**Review by rotemee — CHANGES_REQUESTED** [review]

File: core/editor/editor-loader.php (hunk 13 above)
Comment (Nevoss) [reply core/editor/editor-loader.php:190]: Done

File: core/editor/editor-loader.php [hunk 18]
```diff
@@ -135,7 +183,42 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
+	 * to support loading translations for lazy loaded scripts.
+	 *
+	 * Want to go deeper? Read the following article:
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:193]: Done

File: core/editor/editor-loader.php [hunk 19]
```diff
@@ -22,14 +22,22 @@ public function __construct( Config_Provider_Interface $config_provider ) {
 		$this->config_provider = $config_provider;
 	}
 
+	/**
+	 * @return void
+	 */
+	public function register_actions() {
+		$script_configs = $this->get_script_configs();
+
+		add_filter( 'load_script_textdomain_relative_path', function ( $relative_path, $src ) use ( $script_configs ) {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:31]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 20]
```diff
@@ -0,0 +1,117 @@
+const { sources: { RawSource } } = require( 'webpack' );
+
+const MODULE_FILTER = [ /((?:[^!?\s]+?)(?:\.js|\.jsx|\.ts|\.tsx))$/, /^((?!node_modules).)*$/ ];
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:3]: done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+		this.translationsRegexps = translationsRegexps.map( ( regex ) => {
+			const flags = [ ...new Set( [ 'g', 'm', ...regex.flags.split( '' ) ] ) ].join( '' );
+
+			return new RegExp( regex.source, flags );
+		} );
+	}
+
+	apply( compiler ) {
+		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
+			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:26]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+
+		this.translationsRegexps = translationsRegexps.map( ( regex ) => {
+			const flags = [ ...new Set( [ 'g', 'm', ...regex.flags.split( '' ) ] ) ].join( '' );
+
+			return new RegExp( regex.source, flags );
+		} );
+	}
+
+	apply( compiler ) {
+		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:25]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
+				const translationCallExpressions = this.getTranslationCallExpressions( compilation );
+
+				this.addTranslationCallExpressionsToAssets( compilation, translationCallExpressions );
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:29]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js (hunk 14 above)
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:86]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js (hunk 15 above)
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:59]: Done

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js (hunk 16 above)
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:71]: Done

File: packages/webpack.prod.js (hunk 17 above)
Comment (Nevoss) [reply packages/webpack.prod.js:18]: Done

File: core/editor/config-providers/editor-v2-config-provider.php [hunk 24]
```diff
@@ -14,29 +14,45 @@ public function get_script_configs() {
...
+					'deps' => [ 'react', 'elementor-packages-editor', 'elementor-packages-ui', 'wp-i18n' ],
+					'translations' => [
+						'active' => true,
+						'file_suffix' => 'strings',
+					],
 				],
 				[
 					'handle' => 'elementor-packages-editor',
-					'src' => '{{ASSETS_URL}}/js/packages/editor{{MIN_SUFFIX}}.js',
+					'src' => '{{ASSETS_URL}}js/packages/editor{{MIN_SUFFIX}}.js',
```
Comment (eyalnathan) [inline core/editor/config-providers/editor-v2-config-provider.php:35]: ASSETS_URL -> ELEMENTOR_ASSETS_URL

File: core/editor/editor-loader.php [hunk 25]
```diff
@@ -135,7 +185,45 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
+			} );
+	}
+
+	/**
+	 * The purpose of this function is to point WP to request translations for other file
+	 * instead of the default file for translations.
+	 *
+	 * The idea is that all the translations for specific webpack entry locate in the same .js file
+	 * and not split into multiple files (when using lazy loading the main entry file split into multiple files),
+	 *
```
Comment (eyalnathan) [inline core/editor/editor-loader.php:197]: Not clear, consider re-phrasing.

File: core/editor/editor.php [hunk 26]
```diff
@@ -788,6 +787,7 @@ public function set_post_id( $post_id ) {
 	private function get_loader() {
 		if ( ! $this->loader ) {
 			$this->loader = new Editor_Loader( Config_Provider_Factory::create() );
+			$this->loader->register_actions();
```
Comment (eyalnathan) [inline core/editor/editor.php:790]: change `register_actions()` to a more generic name, e.g. `register_hooks()`

File: tests/phpunit/elementor/core/editor/test-editor-loader.php [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+						'active' => true,
+						'file_suffix' => 'strings',
+					]
+				],
+				[
+					'handle' => 'script-not-minified',
+					'src' => 'http://localhost/assets/js/script-not-minified.js',
+					'translations' => [
+						'active' => true,
+						'file_suffix' => 'strings',
```
Comment (eyalnathan) [inline tests/phpunit/elementor/core/editor/test-editor-loader.php:148]: add case with different suffix

File: core/editor/config-providers/editor-v2-config-provider.php (hunk 24 above)
Comment (Nevoss) [reply core/editor/config-providers/editor-v2-config-provider.php:35]: done

File: core/editor/editor.php (hunk 26 above)
Comment (Nevoss) [reply core/editor/editor.php:790]: Done

File: tests/phpunit/elementor/core/editor/test-editor-loader.php [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+			[
+				'some-dep',
+				'wp-i18n',
+			]
+		);
+
+		// Cleanup
+		wp_deregister_script( 'script-with-translations' );
+		wp_deregister_script( 'script-with-another-domain' );
+		wp_deregister_script( 'script-without-translations' );
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/editor/test-editor-loader.php:126]: Done

File: tests/phpunit/elementor/core/editor/test-editor-loader.php (hunk 27 above)
Comment (Nevoss) [reply tests/phpunit/elementor/core/editor/test-editor-loader.php:148]: Done

File: core/editor/editor-loader.php (hunk 25 above)
Comment (Nevoss) [reply core/editor/editor-loader.php]: Done

**Review by rotemee — CHANGES_REQUESTED** [review]

File: core/editor/editor-loader.php [hunk 29]
```diff
@@ -135,7 +185,45 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
-			} )
-			->all();
+			} );
+	}
+
+	/**
+	 * The purpose of this function is to point WP to request translations for other file
+	 * instead of the default file for translations.
+	 *
+	 * The idea is that all the translations for specific webpack entry locate in the same .js file
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php]: Changed

File: packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+			return new RegExp( regex.source, flags );
+		} );
+	}
+
+	apply( compiler ) {
+		// Learn more about Webpack plugin system: https://webpack.js.org/api/plugins/
+
+		// Learn more about Webpack compilation process and hooks: https://webpack.js.org/api/compilation-hooks/
+		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
+			// We tap into the time Webpack done processing all the other assets.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/tools/webpack/extract-i18n-expressions-webpack-plugin.js:29]: Done

File: core/editor/editor-loader.php [hunk 31]
```diff
@@ -135,7 +196,44 @@ private function normalize_asset_configs( array $script_configs, $type ) {
...
+	 * The purpose of this function is to replace the requested translation file
+	 * with a file that contains all the translations for specific scripts.
+	 *
+	 * When writing a script and using Webpack's lazy load feature,
+	 * the script will be split into multiple chunks,
+	 * therefore the translations that generated by WordPress will also split into multiple files,
+	 * for that reason we replace the requested translation file with another file (generated in the build process)
+	 * that contains all the translations for the specific script (including lazy load chunks).
+	 *
+	 * Want to go deeper? Read the following article:
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/editor/editor-loader.php:212]: done

**Review by rotemee — DISMISSED** [review]

**Review by eyalnathan — APPROVED** [review]
LGTM

**Pull request merged**